source/routing/journey_pattern_container.h

```diff
@@ -0,0 +1,191 @@
+        return jpps[idx.val];
+    }
+    JpRange get_jps() const {
+        return boost::make_iterator_range(JpIterator(0, jps.begin()),
+                                          JpIterator(jps.size(), jps.end()));
+    }
+    const Jps& get_jps_values() const { return jps; }
+    JppRange get_jpps() const {
+        return boost::make_iterator_range(JppIterator(0, jpps.begin()),
+                                          JppIterator(jpps.size(), jpps.end()));
+    }
+    const Jpps& get_jpps_values() const { return jpps; }
+    const IdxMap<type::Route, std::vector<JpIdx>>& get_jps_from_route() const {
+        return jps_from_route;
+    }
+    const IdxMap<type::PhysicalMode, std::vector<JpIdx>>& get_jps_from_phy_mode() const {
+        return jps_from_phy_mode;
+    }
+    const IdxMap<type::VehicleJourney, JpIdx>& get_jp_from_vj() const {
+        return jp_from_vj;
+    }
+    const JppIdx& get_jpp(const type::StopTime&) const;
+    std::string get_id(const JpIdx&) const;
+    std::string get_id(const JppIdx&) const;
+    boost::optional<JpIdx> get_jp_from_id(const std::string&) const;
+    boost::optional<JppIdx> get_jpp_from_id(const std::string&) const;
+
+private:
+    struct JppKey {
+        JppKey(const SpIdx& sp, const uint16_t& ltz, const std::bitset<8> p):
+            sp_idx(sp), local_traffic_zone(ltz), properties(p) {
+        }
+        bool operator<(const JppKey& other) const {
+            if (sp_idx != other.sp_idx) { return sp_idx < other.sp_idx; }
+            if (local_traffic_zone != other.local_traffic_zone) {
+                return local_traffic_zone < other.local_traffic_zone;
+            }
+            return properties.to_ulong() < other.properties.to_ulong();
+        }
+
+        // As stop time, but without departure and arrival time.
+        SpIdx sp_idx;
+        uint16_t local_traffic_zone;
+        std::bitset<8> properties;
+    };
+    struct JpKey {
+        // To be in the same jp, the vjs must have the same stop times
+        // (without the departure and arrival times)
+        std::vector<JppKey> jpp_keys;
+
+        // The route is not needed by raptor, but needed by the other
+        // users of get_stop_times, thus we keep it for the moment
+        // (and in practice, it should change nothing).
+        RouteIdx route_idx;
+
+        // The physical mode is not needed by raptor, but used by
+        // forbidden uris.  To remove it, we may precompute a big
+        // matrix for next_stop_time.  Also, there is ideas of a meta
+        // journey pattern that will not contain the high level
+        // information as the route and the physical mode.
+        PhyModeIdx phy_mode_idx;
+
+        // We do not want a jp with frequency and discrete vj as it is
+        // difficult to check if they overtake.
+        bool is_freq = false;
+
+        bool operator<(const JpKey& other) const {
+            if (route_idx != other.route_idx) { return route_idx < other.route_idx; }
```

Comment:
so route_idx is the most weighted(or important) attribute in the comparison...

Let's say, you have jk1, jk2, jk3, when you compare jk1 with jk2, you base your comparison on route_idx, then, you compare jk2 with jk3, you base the comparison on is_freq, then you compare jk1 and jk3 on phy_mode_idx. 

What can happen is you got jk1 <jk2 and jk2 < jk3, logically you SHOULD get jk1<jk3, but in practice, you may get jk1 >jk3...